toolchain/check/testdata/array/fail_type_mismatch.carbon

```diff
@@ -28,19 +28,24 @@ var d: [i32; 3] = t2;
 
 // CHECK:STDOUT: file "fail_type_mismatch.carbon" {
 // CHECK:STDOUT:   %.loc10_14: i32 = int_literal 3
-// CHECK:STDOUT:   %.loc10_15: type = array_type %.loc10_14, i32
+// CHECK:STDOUT:   %.loc10_15.1: type = array_type %.loc10_14, i32
+// CHECK:STDOUT:   %.loc10_15.2: type = ptr_type [i32; 3]
 // CHECK:STDOUT:   %a: ref [i32; 3] = var "a"
 // CHECK:STDOUT:   %.loc10_20: i32 = int_literal 1
+// CHECK:STDOUT:   %.1: type = ptr_type String
 // CHECK:STDOUT:   %.loc10_23: String = string_literal "Hello"
 // CHECK:STDOUT:   %.loc10_32: String = string_literal "World"
 // CHECK:STDOUT:   %.loc10_39.1: type = tuple_type (i32, String, String)
-// CHECK:STDOUT:   %.loc10_39.2: (i32, String, String) = tuple_literal (%.loc10_20, %.loc10_23, %.loc10_32)
-// CHECK:STDOUT:   %.loc10_39.3: i32 = int_literal 0
-// CHECK:STDOUT:   %.loc10_39.4: ref i32 = array_index %a, %.loc10_39.3
-// CHECK:STDOUT:   %.loc10_39.5: init i32 = initialize_from %.loc10_20 to %.loc10_39.4
+// CHECK:STDOUT:   %.loc10_39.2: type = tuple_type (i32, String*, String*)
+// CHECK:STDOUT:   %.loc10_39.3: type = ptr_type (i32, String*, String*)
```

Comment:
Not sure it makes sense to address in this PR, but I think this shows an issue with the current value representation approach for tuples.

If we take the value representation of each member of the tuple, and one of them uses a pointer value rep, and then the tuple *itself* uses a pointer value rep, we don't want to end up with a pointer-to-a-pointer.

When we're selecting a value rep as a pointer, the pointer is to the *object*, and so we shouldn't have the field value reps here, as we'll go all the way back to the tuple *object* with object representations for all its fields. And that's what we want so we don't end up with double-indirection.

Reply:
We should discuss this when you get back from vacation. There's an interesting design tradeoff here that gets into how we want our value semantics to work. As an example to mull over, consider:
```carbon
fn F(strings: (String, String));

fn G(s: String) {
  F((s, s));
}
```
Here, assume that the object representation of `String` owns a heap buffer, and the value representation is just a pointer and size. And that the tuple value representation is large enough that we want to pass it indirectly. (Eg, add tuple elements until we reach that point.)

Either:
1) The value representation of `(String, String)` is a pointer to storage containing two `String` _value_ representations, meaning that we have a case where the value rep is a pointer, but it's not a pointer to an object rep of the same type. Or...
2) The value representation of `(String, String)` is a pointer to storage containing two `String` _object_ representations, meaning that we need to perform heap allocations and copies in order to call `F` from `G`, where we fundamentally had no need to create any new `String` objects. Or...
3) Something else!

To me, option (1) seems like a better choice than option (2). There's then another choice:

Either:

1a) The value representation of a tuple contains the value representations of its elements, always, leading to the pointers-to-pointers behavior you're observing here. Or...
1b) If an element's value representation is a pointer, then the tuple's value representation stores the pointee directly. That'd mean that we avoid the double indirection, at the cost of making additional copies. For very large, heavily nested tuples, this can result in the creation of a lot more stack objects and copying between them, but for cases like this one, may be a substantial improvement. Or...
1c) We heuristically pick between 1a and 1b based on what we think is the right tradeoff between extra indirections and extra copies. Or...
1d) Something else!

My current thinking is that either (1b) or (1c) is probably a good direction. But I think there's a lot to explore here.

(I'll open an issue to track this too.)